tutorials/JSON_serialization_with_serde.md

```diff
@@ -0,0 +1,197 @@
+Let's include those.
+
+```rust
+pub struct Amount(u64)
+pub struct ScriptBuf(Vec<u8>)
+
+pub struct TxOut {
+    /// The value of the output, in satoshis.
+    pub value: Amount,
+    /// The script which must be satisfied for the output to be spent.
+    pub script_pubkey: ScriptBuf,
+}
+```
+
+Now we can add the `derive` attribute to `TxOut` for `Serialize` and see if this works by calling `serde_json::to_string_pretty` on our `TxOut` struct instance.
+Remember to bring `serde_json` and `serde::Serialize` into scope.
+You can follow along on [Rust Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a76739816ac2a17b268a58064dde24b6).
+
+```rust
+use serde_json;
+use serde::Serialize;
+
+pub struct Amount(pub u64);
+pub struct ScriptBuf(pub Vec<u8>);
+
+#[derive(Serialize)]
+pub struct TxOut {
+    /// The value of the output, in satoshis.
+    pub value: Amount,
+    /// The script which must be satisfied for the output to be spent.
+    pub script_pubkey: ScriptBuf,
+}
+
+fn main() -> () {
+    let amount = Amount(1_000_000);
+    let script_pubkey = ScriptBuf(vec![0, 0, 0, 0]);
+
+    let txout = TxOut {
+        amount,
+        script_pubkey,
+    };
+
+    let json_txout = serde_json::to_string_pretty(&txout).unwrap();
+
+    println!("{}", json_txout);
+}
+```
+
+This, of course, won't work. We need to implement the `Serialize` trait for both the `ScriptBuf` and `Amount`. 
+This trait is already implemented by default for most primitive and standard types. 
+However, since these are custom types we'll need to implement the `Serialize` trait for each of them. 
+
+# Problem Statement
+
+Implement the `Serialize` trait for the `Amount` and `ScriptBuf` structs.
+
+## Implementation
+
+When implementing the `Serialize` trait, we must implement the required method, `serialize`.
+This is the method signature:
+`fn serialize<S: Serializer>(&self, s: S) -> Result<S::Ok, S::Error>`.
+It accepts a type which implements the `Serializer` trait.
+This type will be passed in by `serde_json` when we call the `to_string_pretty` method.
+The `Serializer` trait has several methods available that we can use to serialize various types, which can be found in the [documentation](https://docs.rs/serde/latest/serde/ser/trait.Serializer.html).
+
+Let's start with a simple version, leveraging the methods available for the `Serializer` trait.
+
+```rust
+...
+
+impl Serialize for Amount {
+    fn serialize<S: Serializer>(&self, s: S) -> Result<S::Ok, S::Error> {
+        s.serialize_u64(self.0)
+    }
+}
+
+impl Serialize for ScriptBuf {
+    fn serialize<S: Serializer>(&self, s: S) -> Result<S::Ok, S::Error> {
+        s.serialize_bytes(&self.0)
+    }
+}
+
+...
+```
+
+Notice that we use the `serialize_u64` and `serialize_bytes` methods. 
+The `serialize_bytes` method accepts a `slice` as an argument and not a `Vec` so we can append the `&` to `self` to get a slice reference.
```

Comment:
It would be better to say we are using `AsRef` trait rather than saying append `&` to `self`.

Reply:
Yes you are correct, `AsRef` is explicit behavior.
Like you suggested simply using `.as_slice()` will be good!